package-recipe.el

```diff
@@ -47,6 +47,8 @@
    (files           :initarg :files          :initform nil)
    (branch          :initarg :branch         :initform nil)
    (commit          :initarg :commit         :initform nil)
+   (time                                     :initform nil)
```

Comment:
> This change seems like we're mixing abstractions. `package-recipe` to me is a representation of the actual recipe.

Change your mental modal to "build instructions". We start out with the recipe file, which explains how any version of that package has to be build. We then determine which version we want to build. This becomes part of the instructions the code that actually does the building has to follow.

Previously we passed this additional information along as separate arguments. As noted in the commit message this complicated function signatures, and also led to some information not being passed along and instead being recalculated later on. Sticking to separate arguments also means that any change to what information is being passed along, is a breaking change. In the past such changes did affect public functions, going forward it shouldn't happen anymore.

> Is this something we should work towards? I.e., it would make sense that in the future we have a function that takes a recipe as input and returns a snapshot. Both of which are not mutable.

This feels like an unnecessary complication. There isn't just one recipe class but one for each type of upstream repository/forge that we support. We would have to add one "snapshot" class for each "recipe" class. Adding two slots that do not appear in recipe files to the base recipe class, seems like a much cleaner way of implementing this and I don't agree that doing that means that we are mixing abstractions.